### PR TITLE
grammar: on..to statement

### PR DESCRIPTION
**Pull request by kalikiana**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----
This is the implementation of the "to" statement [as discussed in the forum](https://forum.snapcraft.io/t/specifying-architecture-on-cross-compilations/1641/20). This PR implements the **bold items**.
Items 4, 5 and 6 are implemented in #1639.

1. The existing “on <arch>:” syntax will be accepted under build-packages as well
2. The default architecture for packages under “on <arch>” is the one from the build machine
3. Packages under “on <arch>:” are installed only when the build machine has that architecture
4. A new “to <arch>:” statement will be supported in build-packages and stage-packages
5. The default architecture for packages under “to <arch>:” is the target architecture
6. Packages under “to <arch>:” are installed only when the target architecture is the one listed, which includes both a cross-compilation towards that architecture and a native build on that architecture
7. **The two statements may be used together as “on <arch> to <arch>:” to precisely specify the case of a particular cross-compilation**
8. **The default architecture for packages under “on <arch> to <arch>:” is the target architecture**

Docs: [snappy-docs#325](https://github.com/canonical-docs/snappy-docs/pull/325)

### CODE REVIEW

**Comment by kalikiana** [comment]
@kyrofa  I sorted out the test failure and added integration tests for both build- and stage-packages.

File: snapcraft/internal/project_loader/grammar/_processor.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 from .errors import GrammarSyntaxError
 
 _ON_CLAUSE_PATTERN = re.compile(r'\Aon\s+')
-_TO_CLAUSE_PATTERN = re.compile(r'\Ato\s+')
+_TO_CLAUSE_PATTERN = re.compile(r'(\Aon\s+\S+\s+|)to\s+')
```
Comment (kyrofa) [inline snapcraft/internal/project_loader/grammar/_processor.py]: Does this not preclude its use standalone? It seems to break items 4, 5, and 6 of [this list](https://forum.snapcraft.io/t/specifying-architecture-on-cross-compilations/1641/20?u=kyrofa).

Comment (kyrofa) [inline snapcraft/internal/project_loader/grammar/_processor.py]: Clever use of alternation. Got me for a sec.

File: snapcraft/internal/project_loader/grammar/_to.py [hunk 2]
```diff
@@ -146,4 +157,6 @@ def _extract_to_clause_selectors(to):
         raise ToStatementSyntaxError(
             to, message='spaces are not allowed in the selectors')
 
-    return {selector.strip() for selector in selector_group.split(',')}
+    return (
+        {selector.strip() for selector in on_selector_group[3:].split(',')},
```
Comment (kyrofa) [inline snapcraft/internal/project_loader/grammar/_to.py]: This `3` is yucky. I know it's for the `on\s` but I'll probably forget. Would you mind adding a comment for that, please? Also, returning a set containing an empty string for a statement that's completely missing does not seem to be a clear API to me. I would expect it to return e.g. None. Thoughts?

File: snapcraft/tests/integration/general/test_build_package_grammar.py [hunk 3]
```diff
@@ -130,6 +130,16 @@ def test_to_other_arch(self):
         self.run_snapcraft(['pull'])
         self.assertFalse(self._hello_is_installed())
 
+    def test_on_to_other_arch(self):
+        """Test that 'on to' fetches nothing when building for another arch."""
+
+        with self.modified_yaml() as yaml:
+            yaml['parts']['my-part']['build-packages'] = [
+                OrderedDict({'on i386 to other-arch': ['hello']})
```
Comment (kyrofa) [inline snapcraft/tests/integration/general/test_build_package_grammar.py]: I have similar misgivings on this PR's integration tests to the ones I have on #1639.

**Comment by kyrofa** [comment]
As of #1857, the `on` statement behaves the way we need it to in order to ensure snippets A and B work the same way. Rather than essentially writing a new "on" statement here, I'd like to see if we can dynamically rewrite the inline `on..to` to the nested `on: to` by using composition when parsing the grammar. Less duplication that way.

**Comment by sergiusens** [comment]
This closes #1689

**Comment by kyrofa** [comment]
#1857 has been merged, this one is ready for a revamp.

**Comment by kalikiana** [comment]
- The code is refactored now so ToStatement takes an OnStatement
- The `matches` methods can be called from subclasses
- the _SELECTOR_PATTERN of both on and to need to accomodate a combined on..too expression

Caveats to be sorted out:

- I failed to accept illegal whitespace in the _SELECTOR_PATTERN for ToStatement which we want to show in a separate error message
- `on amd64 to amd64` or `on amd64.......to amd64` will be considered duplicates now, and therefore errors. We need to decide how to handle that

**Pull request closed without merging**